### PR TITLE
Store build files in cache

### LINKED ISSUE #179: Store build files in S3 cache

### PR DESCRIPTION
**Pull request by HeDo88TH**
Closes #179 
This PR inherits from #181, do not merge before 👍

### CODE REVIEW

**Comment by HeDo88TH** [comment]
Now the build files are stored alongside the dataset in a .build folder.
Once generated after build, they are synced by a background worker.

**Comment by HeDo88TH** [comment]
CI keeps failing on Github but not on my machines...

**Comment by pierotofy** [comment]
```
System.IO.IOException : The process cannot access the file 
```

Now that we are actually deleting the files properly (due to https://github.com/DroneDB/Registry/commit/e6b07d8e00cbc6ed0d67da7d1f80306d1a1dbf4b) some issues might be arising due to improper assumptions about files being/not being deleted.

File: Registry.Web/Services/Managers/ObjectsManager.cs [hunk 1]
```diff
@@ -46,6 +47,7 @@ public class ObjectsManager : IObjectsManager
 
         // TODO: Could be moved to config
         private const int DefaultThumbnailSize = 512;
+        private const string BuildFolderName = ".build";
```
Comment (pierotofy) [inline Registry.Web/Services/Managers/ObjectsManager.cs:50]: Could this be stored in `.ddb/build` (`<bucket>/.ddb/build`)?

It keeps the FS structure consistent with the DDB folder structure.

File: Registry.Web/Utilities/HangfireUtils.cs [hunk 2]
```diff
@@ -55,7 +57,74 @@ public static void SafeDelete(string path, PerformContext context)
...
+                var source = Path.GetFullPath(Path.Combine(sourcePath, name));
+                var dest = CommonUtils.SafeCombine(destPath, name);
+
+                var contentType = MimeTypes.GetMimeType(file);
+
+#if DEBUG
+                writeLine($"'{file}' -> PutObjectAsync('{name}', '{source}', '{dest}', '{contentType}')");
+#endif
+                // TODO: We could find a way to parallelize this (I know what you are thinking and the answer is NO, more clever plz)
+                objectSystem.PutObjectAsync(bucketName, dest, source, contentType).GetAwaiter().GetResult();
```
Comment (pierotofy) [inline Registry.Web/Utilities/HangfireUtils.cs:107]: lol, no need to optimize at this point, I'm more interested in graceful error handling; I don't see any. What happens if there's a write error on the S3 endpoint?

File: Registry.Web/Services/Managers/ObjectsManager.cs (hunk 1 above)
Comment (HeDo88TH) [reply Registry.Web/Services/Managers/ObjectsManager.cs]: Fixed in https://github.com/DroneDB/Registry/pull/185/commits/9a6b0e873ac4f409c1beb271e890d337515f4baa

File: Registry.Web/Utilities/HangfireUtils.cs (hunk 2 above)
Comment (HeDo88TH) [reply Registry.Web/Utilities/HangfireUtils.cs]: Fixed in https://github.com/DroneDB/Registry/pull/185/commits/9a6b0e873ac4f409c1beb271e890d337515f4baa

**Comment by HeDo88TH** [comment]
Prolly ready @pierotofy

**Comment by pierotofy** [comment]
Did you try building files in subfolders? I can't seem to get them to work.

```
$ ddb init
$ tree .
.
└── abc
    └── tbays.laz

1 directory, 1 file
$ ddb add abc
$ ddb tag http://localhost:5000/admin/test
$ ddb push
```

**Comment by pierotofy** [comment]
There seems to be some severe performance/locking problem happening here.

Try to upload a dataset (upload a LAZ). Things work OK. I can open the 3D view in Hub after a while.

Try to upload a second LAZ file. I can no longer open the **first** LAZ file in the 3D view.

**Comment by pierotofy** [comment]
After a long time, something finishes and I can once again browse both files in 3D.

**Comment by HeDo88TH** [comment]
> There seems to be some severe performance/locking problem happening here.
> 
> Try to upload a dataset (upload a LAZ). Things work OK. I can open the 3D view in Hub after a while.
> 
> Try to upload a second LAZ file. I can no longer open the **first** LAZ file in the 3D view.

> After a long time, something finishes and I can once again browse both files in 3D.

Are the LAZ files big?

**Comment by pierotofy** [comment]
Not really, just ~5MB.

**Comment by pierotofy** [comment]
I would imagine this gets worse the bigger the files are.

**Comment by HeDo88TH** [comment]
> Not really, just ~5MB.

Are you using the cached S3 or just S3?

**Comment by HeDo88TH** [comment]
I have just tried to reproduce this issue but without success. I have to see some logs, maybe something is failing the build phase but It is just my speculation.

**Comment by pierotofy** [comment]
Try running this on a Linux machine perhaps?

**Comment by HeDo88TH** [comment]
> Try running Registry on a Linux machine perhaps?

My linux setup:

- `MINIO_ROOT_USER=minioadmin MINIO_ROOT_PASSWORD=minioadmin ./minio server /home/hedo/data --console-address ":9001"`
- MariaDB: `10.3.31-MariaDB-0ubuntu0.20.04.1 - Ubuntu 20.04`
- S3 without cache

With 16cores and 32GB ram it works flawlessly.

I tried choking it with 2cores and 4GB ram and it froze after uploading a 3x 200+ MB of point clouds.
I think this kind of computations plus chrome with potree made the kill.

**Comment by HeDo88TH** [comment]
> There seems to be some severe performance/locking problem happening here.
> 
> Try to upload a dataset (upload a LAZ). Things work OK. I can open the 3D view in Hub after a while.
> 
> Try to upload a second LAZ file. I can no longer open the **first** LAZ file in the 3D view.

I think there is a bug on hub that duplicates the uploaded files by creating a "shadow". If you try to open the 3D tab of the wrong file it won't work. After refresh it is ok. We need to investigate this.

**Comment by pierotofy** [comment]
You're right, something funky is going on with the 3D tab.

**Pull request merged**